tests/csapi/room_messages_test.go

```diff
@@ -0,0 +1,222 @@
+	queryParams := url.Values{}
+	queryParams.Set("dir", "f")
+	queryParams.Set("from", token)
+	res := alice.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, client.WithQueries(queryParams))
+	must.MatchResponse(t, res, match.HTTPResponse{
+		StatusCode: http.StatusOK,
+		JSON: []match.JSON{
+			match.JSONArrayEach("chunk", func(r gjson.Result) error {
+				if r.Get("type").Str == "m.room.message" && r.Get("content").Get("body").Str == testMessage {
+					return nil
+				}
+				return fmt.Errorf("did not find correct event")
+			}),
+		},
+	})
+}
+
+// sytest: PUT /rooms/:room_id/send/:event_type/:txn_id sends a message
+// sytest: PUT /rooms/:room_id/send/:event_type/:txn_id deduplicates the same txn id
+func TestSendMessageWithTxn(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{})
+
+	const txnID = "lorem"
+
+	res := alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "r0", "rooms", roomID, "send", "m.room.message", txnID}, client.WithJSONBody(t, map[string]interface{}{
+		"msgtype": "m.text",
+		"body":    "test",
+	}))
+	eventID := client.GetJSONFieldStr(t, client.ParseJSON(t, res), "event_id")
+
+	res = alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "r0", "rooms", roomID, "send", "m.room.message", txnID}, client.WithJSONBody(t, map[string]interface{}{
+		"msgtype": "m.text",
+		"body":    "test",
+	}))
+	eventID2 := client.GetJSONFieldStr(t, client.ParseJSON(t, res), "event_id")
+
+	if eventID != eventID2 {
+		t.Fatalf("two /send requests with same txn ID resulted in different event IDs")
+	}
+}
+
+func TestRoomMessagesLazyLoading(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite) // FIXME: https://github.com/matrix-org/dendrite/issues/2257
+
+	deployment := Deploy(t, b.MustValidate(b.Blueprint{
+		Name: "alice_bob_and_charlie",
+		Homeservers: []b.Homeserver{
+			{
+				Name: "hs1",
+				Users: []b.User{
+					{
+						Localpart:   "@alice",
+						DisplayName: "Alice",
+					},
+					{
+						Localpart:   "@bob",
+						DisplayName: "Bob",
+					},
+					{
+						Localpart:   "@charlie",
+						DisplayName: "Charlie",
+					},
+				},
+			},
+		},
+	}))
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+	charlie := deployment.Client(t, "hs1", "@charlie:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	bob.JoinRoom(t, roomID, nil)
+	charlie.JoinRoom(t, roomID, nil)
+
+	bob.SendEventSynced(t, roomID, b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "test",
+		},
+	})
+
+	beforeToken := alice.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID), client.SyncJoinedTo(alice.UserID, roomID))
```

Comment:
Presumably you mean to wait until bob and charlie have joined, not bob and alice.